### PR TITLE
added basic ArchToken interface, contract and test

### PR DESCRIPTION
**Pull request by jasonstockman**
Live review done by: **Yotam**

Added basic interface, contract and tests for Arch Token

### CODE REVIEW

File: contracts/ArchToken.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/// @dev This is the contract for the Archimedes rewards token
+contract ArchToken is ERC20 {
+    
+    constructor() ERC20("Archimedes Token", "ARCH") {}
+
+    /// @notice Mints tokens to a recipient.
+    /// @param to the account to mint tokens to.
+    /// @param amount the amount of tokens to mint.
+    function mint(address to, uint256 amount) external {
+        _mint(to, amount);
```
Comment (tomerXarchimedes) [inline contracts/ArchToken.sol:16]: Following up on our discussion (after you submitted the PR request - so it's not pair I get it :-) ) 
We are going to start with not-mintable but burnable 

That probably means that
* The `constructor` should probably get a `Treasury` address, mint 100M tokens and send them to it
* We probably should describe it in the contract comments above
* `mint()` function should explicitly revert with an error message
* `burn()` needs to be implemented (and call the internal `_burn`). Also consider emitting an event when we burn

File: test/ArchToken.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    beforeEach(async function () {
+        contract = await ethers.getContractFactory("ArchToken");
+        [owner, user1, user2, ...users] = await ethers.getSigners();
+        token = await contract.deploy();
+        await token.mint(owner.address, tokenSupply);
+    });
+
+    describe("Minting", function () {
+        it("Should be able to mint to specific address", async function () {
```
Comment (tomerXarchimedes) [inline test/ArchToken.js:25]: Based on the change above `minting` should fail

Also we probably want to add `burn` test

File: test/ArchToken.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    describe("Transactions", function () {
+        it("Should not be able to transfer more than total supply", async function () {
+            let amount = ethers.utils.parseUnits("101");
+            await expect(
+                token.transfer(user1.address, amount)
+            ).to.be.revertedWith("ERC20: transfer amount exceeds balance");
+        });
+
+        it("Should be able to transfer tokens between accounts", async function () {
+            let amount = ethers.utils.parseUnits("1");
```
Comment (tomerXarchimedes) [inline test/ArchToken.js:40]: With the pre-mint change - this one becomes a bit more tricky. 

I would also recommend testing calling transfer from accounts that are not an owner or the Treasury (look for `connect`) that means we'll also test that `approve` works

File: test/ArchToken.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    let token;
+    let owner;
+    let user1;
+    let user2;
+    let users;
+
+    beforeEach(async function () {
+        contract = await ethers.getContractFactory("ArchToken");
+        [owner, user1, user2, ...users] = await ethers.getSigners();
+        token = await contract.deploy();
```
Comment (tomerXarchimedes) [inline test/ArchToken.js:20]: Worth checking if pre-mint worked and minted the correct amount in the `constructor`

**Review by tomerXarchimedes — CHANGES_REQUESTED** [review]
Based on the discussion we had (after the PR) - looks like we need to make ARCH not mintable (100M pre-minted in the constructor) and burnable

File: test/ArchToken.js (hunk 3 above)
Comment (jasonstockman) [reply test/ArchToken.js:41]: Will add "transferFrom" functionality and associated tests to check this

File: test/ArchToken.js (hunk 4 above)
Comment (jasonstockman) [reply test/ArchToken.js:18]: Added

File: contracts/ArchToken.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+/// @title Archimedes Token
+/// @dev This is the contract for the Archimedes rewards token
+contract ArchToken is ERC20 {
+    // TODO set treasury address
+    address addressTreasury = msg.sender;
```
Comment (tomerXarchimedes) [inline contracts/ArchToken.sol:10]: I don't think we need that

We can transfer the treasury address with the constructor.
It is a one time thing, I don't see why we need a state or force deploying the contract from the Treasury address

File: test/ArchToken.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        totalSupply = await token.totalSupply();
+    });
+
+    describe("Pre-Mint", function () {
+        it("Should have pre-mint totalSupply of 100m", async function () {
+            // convert from BigNumber to readable value
+            totalSupply = ethers.utils.formatUnits(totalSupply, "ether");
+            // formatUnits() returns a number with the tenths place included
+            expect(totalSupply).to.eq("100000000.0");
```
Comment (tomerXarchimedes) [inline test/ArchToken.js:27]: You probably also want to check `balanceOf` the target address that got 100M ARCH tokens

File: test/ArchToken.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        });
+    });
+
+    // basic end-to-end testing of underlying erc20
+    describe("Transactions", function () {
+        it("Should not be able to transfer() more than total supply", async function () {
+            let amount = totalSupply + 1;
```
Comment (tomerXarchimedes) [inline test/ArchToken.js:34]: There is a lower boundary here. It actually reverts when you try to transfer more than `balanceOf`

It worth checking though that `balanceOf` treasury is less or equal to `totalSupply`

**Review by tomerXarchimedes — CHANGES_REQUESTED** [review]

File: contracts/ArchToken.sol (hunk 1 above)
Comment (jasonstockman) [reply contracts/ArchToken.sol]: There is not external `mint()` by default - just the internal `_mint`, so as long as we don't add one, it won't be available.

Comment (jasonstockman) [reply contracts/ArchToken.sol]: Added a test to make sure `mint()` throws an error "not a function"

Comment (jasonstockman) [reply contracts/ArchToken.sol]: Marking as resolved as we will handle burning in separate PR

**Pull request merged**